### ISSUE TITLE
(Dynamic Routing) Router get stuck in pending state randomly

### ISSUE THREAD

**Issue by TimovanderKamp** [fulcro-legacy/fulcro-incubator#37]
I'm currently using dynamic routing (0.0.35) with nested routers. Both layers of routers have deferred-route targets that trigger a load and use target-ready as a post-mutation.

At random the top router will be stuck in pending state. Yesterday I tried all day to create a workspace where this behaviour is consistently happening, but I failed to do so, and can not tell what causes this issue.

It seems to be a timing issue, something that fulcro 3.0 will resolve in the future.

**Comment by TimovanderKamp** [comment]
This is what is being logged when it will get stuck. I route to the segment ["content" "2" "info"] Where the segment ["content" :content/id] is handled by `RootRouter` and ["info"] is handled by the `ContentRouter`

At some moment when everything should be done, the RootRouter will start a deferred-route to path nil, which will never be notified to be ready.

```
[723.862s] [fulcro.incubator.ui-state-machines] Triggering :route! on :fulcro.incubator.routing-broken-ws/RootRouter with {:error-timeout 5000, :deferred-timeout 100, :path-segment ["content" "2"], :router [:fulcro.incubator.dynamic-routing/id :fulcro.incubator.routing-broken-ws/RootRouter], :target [:content/id 2]} 
console.js:218  [723.863s] [fulcro.incubator.ui-state-machines] Triggering :route! on :fulcro.incubator.routing-broken-ws/ContentRouter with {:error-timeout 5000, :deferred-timeout 100, :path-segment ["info"], :router [:fulcro.incubator.dynamic-routing/id :fulcro.incubator.routing-broken-ws/ContentRouter], :target [:content :info]} 
console.js:218  [723.865s] [fulcro.incubator.ui-state-machines] Storing :pending-path-segment -> ["content" "2"] on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [723.866s] [fulcro.incubator.ui-state-machines] Applying mutation helper to state of :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [723.869s] [fulcro.incubator.ui-state-machines] Activating state  :deferred on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [723.870s] [fulcro.incubator.ui-state-machines] Storing :target -> [:content/id 2] on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [723.871s] [fulcro.incubator.ui-state-machines] Setting timeout :error-timer on :fulcro.incubator.routing-broken-ws/RootRouter to send :timeout! in 5000 ms 
console.js:218  [723.873s] [fulcro.incubator.ui-state-machines] Setting timeout :delay-timer on :fulcro.incubator.routing-broken-ws/RootRouter to send :waiting! in 100 ms 
console.js:218  [723.886s] [fulcro.incubator.ui-state-machines] Storing :pending-path-segment -> ["info"] on :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.887s] [fulcro.incubator.ui-state-machines] Applying mutation helper to state of :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.890s] [fulcro.incubator.ui-state-machines] Activating state  :deferred on :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.892s] [fulcro.incubator.ui-state-machines] Storing :target -> [:content :info] on :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.893s] [fulcro.incubator.ui-state-machines] Setting timeout :error-timer on :fulcro.incubator.routing-broken-ws/ContentRouter to send :timeout! in 5000 ms 
console.js:218  [723.895s] [fulcro.incubator.ui-state-machines] Setting timeout :delay-timer on :fulcro.incubator.routing-broken-ws/ContentRouter to send :waiting! in 100 ms 
console.js:218  [723.972s] [fulcro.incubator.dynamic-routing] Router :fulcro.incubator.routing-broken-ws/ContentRouter notified that pending route is ready. 
console.js:218  [723.972s] [fulcro.incubator.ui-state-machines] Triggering :ready! on :fulcro.incubator.routing-broken-ws/ContentRouter with {} 
console.js:218  [723.975s] [fulcro.incubator.dynamic-routing] Router :fulcro.incubator.routing-broken-ws/RootRouter notified that pending route is ready. 
console.js:218  [723.976s] [fulcro.incubator.ui-state-machines] Triggering :ready! on :fulcro.incubator.routing-broken-ws/RootRouter with {} 
console.js:218  [723.978s] [fulcro.incubator.ui-state-machines] Storing :path-segment -> ["info"] on :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.980s] [fulcro.incubator.ui-state-machines] Storing :pending-path-segment -> [] on :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.982s] [fulcro.incubator.ui-state-machines] Applying mutation helper to state of :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.983s] [fulcro.incubator.dynamic-routing] Applying route ident [:content :info] to router :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.985s] [fulcro.incubator.ui-state-machines] Activating state  :routed on :fulcro.incubator.routing-broken-ws/ContentRouter 
console.js:218  [723.987s] [fulcro.incubator.ui-state-machines] Cancelling timer  :error-timer on :fulcro.incubator.routing-broken-ws/ContentRouter due to event :ready! 
console.js:218  [723.988s] [fulcro.incubator.ui-state-machines] Clearing timeout  :fulcro.incubator.routing-broken-ws/ContentRouter : :error-timer 
console.js:218  [723.990s] [fulcro.incubator.ui-state-machines] Cancelling timer  :delay-timer on :fulcro.incubator.routing-broken-ws/ContentRouter due to event :ready! 
console.js:218  [723.992s] [fulcro.incubator.ui-state-machines] Clearing timeout  :fulcro.incubator.routing-broken-ws/ContentRouter : :delay-timer 
console.js:218  [723.995s] [fulcro.incubator.ui-state-machines] TIMEOUT on :fulcro.incubator.routing-broken-ws/RootRouter due to timer :delay-timer after 100 ms 
console.js:218  [723.996s] [fulcro.incubator.ui-state-machines] Triggering :waiting! on :fulcro.incubator.routing-broken-ws/RootRouter with {} 
console.js:218  [724.008s] [fulcro.incubator.ui-state-machines] Storing :path-segment -> ["content" "2"] on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.009s] [fulcro.incubator.ui-state-machines] Storing :pending-path-segment -> [] on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.011s] [fulcro.incubator.ui-state-machines] Applying mutation helper to state of :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.012s] [fulcro.incubator.dynamic-routing] Applying route ident [:content/id 2] to router :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.014s] [fulcro.incubator.ui-state-machines] Activating state  :routed on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.016s] [fulcro.incubator.ui-state-machines] Cancelling timer  :error-timer on :fulcro.incubator.routing-broken-ws/RootRouter due to event :ready! 
console.js:218  [724.017s] [fulcro.incubator.ui-state-machines] Clearing timeout  :fulcro.incubator.routing-broken-ws/RootRouter : :error-timer 
console.js:218  [724.019s] [fulcro.incubator.ui-state-machines] Cancelling timer  :delay-timer on :fulcro.incubator.routing-broken-ws/RootRouter due to event :ready! 
console.js:218  [724.020s] [fulcro.incubator.ui-state-machines] Clearing timeout  :fulcro.incubator.routing-broken-ws/RootRouter : :delay-timer 
***** ROOT ROUTER STARTS TO ROUTE TO nil ****** console.js:218 * [724.024s] [fulcro.incubator.ui-state-machines] Storing :pending-path-segment ->  on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.026s] [fulcro.incubator.ui-state-machines] Applying mutation helper to state of :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.030s] [fulcro.incubator.ui-state-machines] Activating state  :deferred on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.032s] [fulcro.incubator.ui-state-machines] Storing :target ->  on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [724.034s] [fulcro.incubator.ui-state-machines] Setting timeout :error-timer on :fulcro.incubator.routing-broken-ws/RootRouter to send :timeout! in 5000 ms 
console.js:218  [724.036s] [fulcro.incubator.ui-state-machines] Setting timeout :delay-timer on :fulcro.incubator.routing-broken-ws/RootRouter to send :waiting! in 100 ms 
console.js:218  [724.139s] [fulcro.incubator.ui-state-machines] TIMEOUT on :fulcro.incubator.routing-broken-ws/RootRouter due to timer :delay-timer after 100 ms 
console.js:218  [724.140s] [fulcro.incubator.ui-state-machines] Triggering :waiting! on :fulcro.incubator.routing-broken-ws/RootRouter with {} 
console.js:218  [724.149s] [fulcro.incubator.ui-state-machines] Activating state  :pending on :fulcro.incubator.routing-broken-ws/RootRouter 
console.js:218  [729.041s] [fulcro.incubator.ui-state-machines] TIMEOUT on :fulcro.incubator.routing-broken-ws/RootRouter due to timer :error-timer after 5000 ms 
console.js:218  [729.044s] [fulcro.incubator.ui-state-machines] Triggering :timeout! on :fulcro.incubator.routing-broken-ws/RootRouter with {} 
console.js:218  [729.053s] [fulcro.incubator.ui-state-machines] Activating state  :failed on :fulcro.incubator.routing-broken-ws/RootRouter 
```

**Comment by awkay** [comment]
So, this patch was an attempt to fix those for the time being:

https://github.com/fulcrologic/fulcro-incubator/commit/4703a75c3ceb65c5ae89624fc85583d63b9a0abc

but you're using it.  That is the exact area of code you're hitting.  You might try using :local/root with deps to pull incubator into your project and play with adding some more logging around that area.  Perhaps you can figure out the condition.

**Comment by awkay** [comment]
Actually, I think this might be something else. The state machine is triggering a timeout as if you've missed a deferred routing deadline:

console.js:218  [729.041s] [fulcro.incubator.ui-state-machines] TIMEOUT on :fulcro.incubator.routing-broken-ws/RootRouter due to timer :error-timer after 5000 ms 

which either means you forgot to resolve the route *OR* you managed to resolve it before the state machine had moved to the next state that could recognize the event and the timer did not get cancelled.  That is causing the route failure timeout to fire.

**Comment by TimovanderKamp** [comment]
Thanks for your reply, I will try to find out what is going on by adding some additional logging

**Comment by awkay** [comment]
I released another patch for this router into incubator yesterday.  COuld be related to the original issue, since the result was the route getting stuck (it had to do with more than one component having the same ident, but only one being a  router target). Released.
